### DISCUSSION TITLE
Jdbc连接池在kyuubi下的最佳实践

### DISCUSSION THREAD

**Discussion by imperio-wxm** [apache/kyuubi#3987] [Q&A]
想用jdbc连接池维护kyuubi的jdbc连接，有没有最佳实践
每个用户利用subdomain绑定多个实例，希望用户在多个subdomain之间负载均衡请求
subdomain参数及其他信息都是通过jdbcurl传入，如果要利用连接池则一个用户需要创建多个DataSource，很不好维护

1. kyuubi是否建议利用连接池，还是每次查询只建立connection单连用完就关闭？
2. 一个用户想要拥有多driver，并且在多driver间负载均衡用subdomain是否是合理选择？

**Comment by yaooqinn** [comment]
> 想用jdbc连接池维护kyuubi的jdbc连接，有没有最佳实践


连接池是可以支持的，不过同时需要注意 Kyuubi 对于连接和后端engine的共享关系。比如sharelevel=connection，在这种模式下，engine是跨连接共享的，这时候的连接池单单表现为占用服务端多个线程。而要实现engine的共享则需要 sharelevel>=user。

> 每个用户利用subdomain绑定多个实例，希望用户在多个subdomain之间负载均衡请求

Kyuubi支持engine pool，这个pool里可以设置engine的个数，无论是客户端连接池方式，还是临时性的连接，都可以对应到一个pool进行负载均衡。
其中，engine pool是基于subdomain实现的。相当于一个subdomain组。

> kyuubi是否建议利用连接池，还是每次查询只建立connection单连用完就关闭？

具体场景问题具体分析吧，连接的保持对于kyuubi 来讲也就是engine的常驻，连接池和engine的常驻等能力可以保证性能最大化，但同时 Spark 等计算框架在常驻模式下的稳定性也是一个重要的挑战

> 一个用户想要拥有多driver，并且在多driver间负载均衡用subdomain是否是合理选择？

driver 是指jdbc driver，以及同个进程还是多个？ 还是spark driver？

**Reply by imperio-wxm** [reply]
driver 是指spark driver，一个用户拥有多个spark driver。
尝试配置kyuubi.engine.pool.size=2，但是只启动了一个spark driver，想要的效果是一个用户拥有多个spark driver，然后在多个spark driver间负载均衡，或者根据不同job类型提交到不同的spark driver上。

**Reply by pan3793** [reply]
> ...但是只启动了一个spark driver...

如何观测到的该现象？Spark 引擎是按需启动的，意味着你必须创建多个 JDBC Connection (每个对应一个 Kyuubi Session)，才能拉起多个 Spark 引擎

**Reply by imperio-wxm** [reply]
如何观测到的该现象？
> spark部署在k8s上，通过k8上同一个用户的driver pod数量判断

我对engine.pool的概念有些疑惑，是一个engine对应一个spark driver + 一组Executor，还是engine是个逻辑的概念一个engine可以对应多个spark driver？

例如这样的url连接：
`jdbc:kyuubi://127.0.0.1:10009/default;user=tom;?kyuubi.engine.pool.size=3`
我的理解此时应该拉起3个spark driver当成engine pool，用户的请求会在3个spark dirver间负载均衡
但实际只启动了一个spark driver，此时spark driver依旧是个单点，engine pool的作用是什么？

配合subdomain的url连接：
`jdbc:kyuubi://127.0.0.1:10009/default;user=tom;?kyuubi.engine.share.level.subdomain=01;kyuubi.engine.pool.size=3`
`jdbc:kyuubi://127.0.0.1:10009/default;user=tom;?kyuubi.engine.share.level.subdomain=02;kyuubi.engine.pool.size=3`
此时subdomain=01会启动一个spark driver，subdomain=02会启动一个spark driver，这个时候意味着同一个tom用户拥有了2个spark driver，这样对于这个用户来说请求可以在subdomain=01、subdomain=02之间负载均衡，做到spark driver的高可用

我有点不太明白kyuubi engine 和 spark driver 之间是什么样的关系？

**Reply by pan3793** [reply]
Kyuubi Engine 即一个 Spark Application，每个 Spark Application 由一个 Driver 和多个 Executor 组成。部分语境中，Kyuubi Engine、Spark App、Spark Driver 可以等价互换

**Reply by pan3793** [reply]
```
val url = "jdbc:kyuubi://127.0.0.1:10009/default;user=tom;?kyuubi.engine.pool.size=3"
val c1 = Driver.getConnection(url) // randomly get or create one engine from pool (size=3)
// at this time, there is one engine launched in the pool
val c2 = Driver.getConnection(url) // randomly get or create one engine from pool (size=3)
// at this time, there is AT MOST two engines launched in the pool
val c3 = Driver.getConnection(url) // randomly get or create one engine from pool (size=3)
// at this time, there is AT MOST three engines launched in the pool
val c4 = Driver.getConnection(url) // randomly get or create one engine from pool (size=3)
// at this time, there is AT MOST three engines launched in the pool
```

**Reply by imperio-wxm** [reply]
感谢解答！

**Reply by imperio-wxm** [reply]
你好，再问下jdbc方式是否支持spark job 的cancel？某些情况下我希望主动cancel掉某个jobGroup下的code或者sql，而不是通过timeout被动kill掉。

**Reply by pan3793** [reply]
可以使用标准 JDBC API cancel statement
